README.md
[![Gem Version](http://img.shields.io/gem/v/jekyll-gitlab-letsencrypt.svg)](https://rubygems.org/gems/jekyll-gitlab-letsencrypt) [![Build Status](http://img.shields.io/travis/JustinAiken/jekyll-gitlab-letsencrypt/master.svg)](http://travis-ci.org/JustinAiken/jekyll-gitlab-letsencrypt) [![Coveralls branch](http://img.shields.io/coveralls/JustinAiken/jekyll-gitlab-letsencrypt/master.svg)](https://coveralls.io/r/JustinAiken/jjekyll-gitlab-letsencrypt?branch=master) [![Code Climate](http://img.shields.io/codeclimate/github/JustinAiken/jekyll-gitlab-letsencrypt.svg)](https://codeclimate.com/github/JustinAiken/jekyll-gitlab-letsencrypt)

# jekyll-gitlab-letsencrypt

![img](doc/image.png)

This plugin automagically does the entire the letsencrypt process for your gitlab-hosted jekyll blog!

- *(automatic)* It registers your email to the letsencrypt server
- *(automatic)* It generates a challenge file, and commits it directly via the gitlab API
- *(automatic)* It sleeps until the challenge file is live on the internet
- *(automatic)* It asks letsencrypt to verify it
- *(automatic)* It spits out the certificate chain and private key
- *(automatic)* It updates the gitlab pages domain settings to use the certificate

## Usage

### Prerequisites

You must have:
- A jekyll blog
  - Hosted on gitlab pages
  - With a domain name set up and working
- Gitlab CI setup such that when you push to `master` (or your preferred branch), your changes are deployed live

Versions supported:
- Jekyll 3+
  - 3.3.x - 3.8.x is tested against
- Ruby 2.1+
  - Although 2.2+ recommend, as Jekyll itself doesn't support 2.1

### Installation

- Add to your Gemfile:

```ruby
  group :jekyll_plugins do
    gem 'jekyll-emojis'
    gem 'jekyll-more-emojis'
++  gem 'jekyll-gitlab-letsencrypt'
  end
```

and run `bundle install`

## First-time Configuration

- Get a personal access token: https://gitlab.com/profile/personal_access_tokens

Add a `gitlab-letsencrypt` to your `_config.yml`:

```yaml
gitlab-letsencrypt:
  # Gitlab settings:
  personal_access_token: 'MUCH SECRET'             # Gotten from the step above ^^
  gitlab_repo:           'gitlab_user/gitlab_repo' # Namespaced repository identifier

  # Domain settings:
  email:                 'example@example.com'     # Let's Encrypt email address
  domain:                'example.com'             # Domain that the cert will be issued for

  # Jekyll settings:
  base_path:  './'               # Where you want the file to go
  pretty_url: false              # Add a "/" on the end of the URL... set to `true` if you use permalink_style: pretty
  append_str: ''                 # Append this string to the end of the challenge URL
  filename:   'letsencrypt.html' # What to call the generated challenge file

  # Delay settings:
  initial_delay: 120 # How long to wait for Gitlab CI to push your changes before it starts checking
  delay_time:     15 # How long to wait between each check once it starts looking for the file

  # Optional settings you probably don't need:
  gitlab_url:     'https://someurl'           # Set if you need to use a self-hosted GitLab instance
  endpoint:       'https://somewhere'         # if you're doing the ACME thing outside of letsencrypt
  branch:         'master'                    # Defaults to master, but you can use a different branch
  layout:         'null'                      # Layout to use for challenge file - defaults to null, but you can change if needed
  scheme:         'https'                     # Scheme to use for challenge request; default http
  commit_message: 'Renew certificate [ROBOT]' # Commit message to use; defaults to "Automated Let's Encrypt renewal"
```

### Running

- Just type `jekyll letsencrypt`

```shell
$ jekyll letsencrypt
Registering example@example.com to https://acme-v01.api.letsencrypt.org/...
Pushing file to Gitlab
Commiting challenge file as lets.html
Done Commiting! Check https://gitlab.com/gitlab_user/gitlab_repo/commits/master
Going to check http://example.com/.well-known/acme-challenge/lots_of_numbers/ for the challenge to be present...
Waiting 120 seconds before we start checking for challenge..
Got response code 404, waiting 15 seconds...
Got response code 404, waiting 15 seconds...
Got response code 200, file is present!
Requesting verification...
Challenge status = valid
Challenge is valid!
Certificate retrieved!
Updating domain example.com pages setting with new certificates..
Success!
```

### Alternative token usage

If you don't want to put your secret gitlab token in your `_config.yml`, you can pass it as an ENV var when you run the command:

```bash
GITLAB_TOKEN="VERY_SECRET_NOT_IN_GIT_PLEASE" jekyll letsencrypt
```

# License

MIT

# Credits/thanks

- :heart: Gitlab for free page hosting, free repos, and free CI!
- :heart: the Jekyll team for the easy-to-use blogging engine!
- Inspired by the excellent [gitlab-letsencrypt](https://github.com/rolodato/gitlab-letsencrypt) npm package.
- Thanks to contributors:
  - @ethernet-zero
  - @daehlith


CHANGELOG.markdown
# 0.4.1
  - #23 - Add option to append arbitrary text to the challenge file

# 0.4.0

  - Updated travis testing matrix
  - #21 - Add commit message configuration option
    - Thanks to @daehlith !

# 0.3.0

  - #17 - Check for repo id fetch request status
  - Update to Gitlab API V4
    - Mostly from #18 - thanks @ethernet-zero !

# 0.2.1

- Add custom GitLab URL setting

# 0.2.0

- #8 - Fully automated! Use new gitlab API for the last step

# 0.1.0

- #6 - Added option to get secret gitlab token from env var

# 0.0.2

- #3 - Added option to specify scheme used for challenge url request

# 0.0.1

- Initial Release


spec/spec_helper.rb
require 'jekyll'
require 'vcr'

unless ENV["NO_COVERALLS"]
  require 'coveralls'
  Coveralls.wear!
end

$LOAD_PATH.unshift File.expand_path("../../lib", __FILE__)
require "jekyll/gitlab/letsencrypt"

VCR.configure do |config|
  config.cassette_library_dir = "spec/fixtures/vcr_cassettes"
  config.hook_into :faraday
  config.allow_http_connections_when_no_cassette = true
end

RSpec.configure do |config|
  config.filter_run :focus
  config.run_all_when_everything_filtered = true

  config.before { Jekyll::Gitlab::Letsencrypt::Configuration.reset! }
end


spec/lib/jekyll/gitlab/letsencrypt/acme_spec.rb
require 'spec_helper'

describe Jekyll::Gitlab::Letsencrypt::Acme do
  let(:acme)     { described_class.new }

  let(:email)    { 'example@example.com' }
  let(:domain)   { 'example.com' }
  let(:endpoint) { 'https://acme-v01.api.letsencrypt.org/' }

  before do
    allow(Jekyll::Gitlab::Letsencrypt::Configuration).to receive(:email).and_return    email
    allow(Jekyll::Gitlab::Letsencrypt::Configuration).to receive(:domain).and_return   domain
    allow(Jekyll::Gitlab::Letsencrypt::Configuration).to receive(:endpoint).and_return endpoint
  end

  describe '#register!' do
    it 'logs, registers, and returns itself' do
      VCR.use_cassette 'authorization' do
        expect(Jekyll.logger).to receive(:info).with 'Registering example@example.com to https://acme-v01.api.letsencrypt.org/...'
        expect(acme.register!).to eq acme
      end
    end
  end

  describe '#authorized?' do
    pending
  end

  describe '#challenge' do
    pending
  end

  describe '#client' do
    it 'returns a new ACME client' do
      expect(acme.client.endpoint).to eq endpoint
    end
  end
end


spec/lib/jekyll/gitlab/letsencrypt/process_spec.rb
require 'spec_helper'

describe Jekyll::Gitlab::Letsencrypt::Process do
  let(:process) { described_class.new client }
  let(:client)  { Jekyll::Gitlab::Letsencrypt::Acme.new }

  describe '#process!' do
    before do
      allow(process).to receive :commit_to_gitlab!
      allow(process).to receive :wait_until_challenge_is_present
      allow(process).to receive :request_verification!
      allow(process).to receive :await_verification_confirmation
      allow(process).to receive :update_gitlab_pages
      allow(process).to receive :display_certificate
    end

    after { process.process! }

    context 'when already authorized' do
      before { allow(client).to receive(:authorized?).and_return true }

      it 'bails early' do
        expect(Jekyll.logger).to receive(:abort_with).with("Client is already authorized.")
      end
    end

    context "when not authorized" do
      before { expect(client).to receive(:authorized?).and_return false }

      pending
    end
  end

  describe '#challenge_url' do
    subject         { process.send :challenge_url }
    let(:challenge) { double 'challenge', filename: 'foo' }
    before          { allow(client).to receive(:challenge).and_return challenge }

    context "by default" do
      it { should match /http\:/ }
    end

    context 'if scheme is overriden' do
      before { allow(Jekyll::Gitlab::Letsencrypt::Configuration).to receive(:scheme).and_return 'https' }
      it     { should match /https\:/ }
    end
  end
end


lib/jekyll/gitlab/letsencrypt/acme.rb
require 'openssl'
require 'acme-client'

module Jekyll
  module Gitlab
    module Letsencrypt
      class Acme

        delegate :email, :endpoint, :domain, to: Configuration

        attr_accessor :registration

        def register!
          Jekyll.logger.info "Registering #{email} to #{endpoint}..."
          @registration = client.register contact: "mailto:#{email}"
          @registration.agree_terms
          self
        end

        def authorized?
          authorization.status == 'valid'
        end

        def challenge
          @challenge ||= authorization.http01
        end

        def client
          @client ||= begin
            private_key = OpenSSL::PKey::RSA.new(4096)
            ::Acme::Client.new private_key: private_key, endpoint: endpoint, connection_options: { request: { open_timeout: 5, timeout: 5 } }
          end
        end

      private

        def authorization
          @authorization ||= client.authorize(domain: domain)
        end
      end
    end
  end
end


lib/jekyll/gitlab/letsencrypt/process.rb
require 'faraday'

module Jekyll
  module Gitlab
    module Letsencrypt
      class Process

        attr_accessor :client

        def self.process!
          client = Acme.new.register!
          self.new(client).process!
        end

        delegate :base_path, :gitlab_url, :gitlab_repo, :pretty_url?, :append_str, :layout, :domain, :initial_delay, :delay_time, :scheme, to: Configuration

        def initialize(client)
          @client = client
        end

        def process!
          Jekyll.logger.abort_with "Client is already authorized." if client.authorized?

          commit_to_gitlab!
          wait_until_challenge_is_present
          request_verification!
          await_verification_confirmation
          if update_gitlab_pages
            Jekyll.logger.info "Success!"
          else
            Jekyll.logger.info "Updating certificate failed... manual steps:"
            display_certificate
          end

          Jekyll.logger.info "All finished!  Don't forget to \`git pull\` in order to bring your local repo up to date with changes this plugin made."
        end

      private

        def commit_to_gitlab!
          Jekyll.logger.info "Pushing file to Gitlab"
          gitlab_client.commit!(challenge_content)
        end

        def wait_until_challenge_is_present
          Jekyll.logger.info "Going to check #{challenge_url} for the challenge to be present..."
          Jekyll.logger.info "Waiting #{initial_delay} seconds before we start checking for challenge.."
          sleep initial_delay

          loop do
            response = Faraday.get challenge_url
            if response.success?
              Jekyll.logger.info "Got response code #{response.status}, file is present!"
              return
            end
            Jekyll.logger.info "Got response code #{response.status}, waiting #{delay_time} seconds..."
            sleep delay_time
          end
        end

        def request_verification!
          Jekyll.logger.info "Requesting verification..."
          challenge.request_verification
        rescue ::Acme::Client::Error::BadNonce
          Jekyll.logger.info "bad nonce! trying again.."
          challenge.request_verification
        end

        def await_verification_confirmation
          tries = 0
          loop do
            tries = tries + 1
            if challenge.authorization.verify_status == 'valid'
              Jekyll.logger.info "Challenge is valid!"
              return
            end
            Jekyll.logger.info "Challenge status = #{challenge.authorization.verify_status}"
            Jekyll.logger.abort_with "Challenge failed to verify" if tries >= 3
            sleep delay_time
          end
        end

        def update_gitlab_pages
          gitlab_client.update_certificate! certificate.fullchain_to_pem, certificate.request.private_key.to_pem
        end

        def display_certificate
          Jekyll.logger.info "Certifcate retrieved!"
          Jekyll.logger.info "Go to #{gitlab_url}/#{gitlab_repo}/pages"
          Jekyll.logger.info " - If you already have an existing entry for #{domain}, remove it"
          Jekyll.logger.info " - Then click + New Domain and enter the following:"
          Jekyll.logger.info ""
          Jekyll.logger.info "Domain: #{domain}"
          Jekyll.logger.info ""
          Jekyll.logger.info "Certificate (PEM): "
          Jekyll.logger.info certificate.fullchain_to_pem
          Jekyll.logger.info "\n"
          Jekyll.logger.info "Key (PEM): "
          Jekyll.logger.info certificate.request.private_key.to_pem
          Jekyll.logger.info ""
          Jekyll.logger.info ""
          Jekyll.logger.info "... hit save, wait a bit, and your new SSL will be live!"
        end

        def challenge_content
          permalink  = ""
          permalink += base_path if base_path
          permalink += challenge.filename
          permalink += "/" if pretty_url?
          permalink += append_str

          content  = "---\n"
          content += "layout: #{layout}\n"
          content += "permalink: #{permalink}\n"
          content += "---\n"
          content += "\n"
          content += challenge.file_content
          content += "\n"

          content
        end

        def challenge_url
          @challenge_url ||= begin
            url  = "#{scheme}://#{domain}/"
            url += challenge.filename
            url += "/" if pretty_url?
            url
          end
        end

        def gitlab_client
          @gitlab_client ||= GitlabClient.new
        end

        def challenge
          @challenge ||= client.challenge
        end

        def certificate
          @certificate ||= begin
            csr = ::Acme::Client::CertificateRequest.new names: Array(domain)
            client.client.new_certificate csr
          end
        end
      end
    end
  end
end


lib/jekyll/gitlab/letsencrypt/committer.rb
require 'faraday'

module Jekyll
  module Gitlab
    module Letsencrypt
      class Commiter

        attr_accessor :content

        delegate :filename, :personal_access_token, :gitlab_repo, :branch, to: Configuration

        def initialize(content)
          @content = content
        end

        def commit!
          create_branch! unless branch_exists?
          commit_file!
        end

        def create_branch!
          Jekyll.logger.info "Creating branch #{branch}.."
          connection.post do |req|
            req.url  "projects/#{repo_id}/repository/branches"
            req.body = {
              branch_name:  branch,
              ref:          'master'
            }.to_json
          end
        end

        def commit_file!
          Jekyll.logger.info "Commiting challenge file as #{filename}"
          connection.run_request(request_method, nil, nil, nil) do |req|
            req.url        "projects/#{repo_id}/repository/files"
            req.body = {
              file_path:      filename,
              commit_message: "Automated Let's Encrypt renewal",
              branch_name:    branch,
              content:        content
            }.to_json
          end
          Jekyll.logger.info "Done Commiting! Check https://gitlab.com/#{gitlab_repo}/commits/#{branch}"
        end

      private

        def branch_exists?
          response = connection.get "projects/#{repo_id}/repository/branches"
          JSON.parse(response.body).any? { |json| json['name'] == branch }
        end

        def request_method
          response = connection.get "projects/#{repo_id}/repository/files?ref=#{branch}&file_path=#{filename}"
          response.status == 404 ? :post : :put
        end

        def repo_id
          @repo_id ||= begin
            repo_name = gitlab_repo.gsub "/", "%2f"
            response  = connection.get "projects/#{repo_name}"
            JSON.parse(response.body)['id']
          end
        end

        def connection
          @connection ||= Faraday.new(url: 'https://gitlab.com/api/v3/') do |faraday|
            faraday.adapter Faraday.default_adapter
            faraday.headers['Content-Type']  = 'application/json'
            faraday.headers['PRIVATE-TOKEN'] = personal_access_token
          end
        end
      end
    end
  end
end


lib/jekyll/gitlab/letsencrypt/gitlab_client.rb
require 'faraday'

module Jekyll
  module Gitlab
    module Letsencrypt
      class GitlabClient

        attr_accessor :content

        delegate :filename, :personal_access_token, :gitlab_url, :gitlab_repo, :branch, :domain, :commit_message, to: Configuration

        def commit!(content)
          @content = content
          create_branch! unless branch_exists?
          commit_file!
        end

        def update_certificate!(certificate, key)
          Jekyll.logger.info "Updating domain #{domain} pages setting with new certificates.."
          response = connection.put do |req|
            req.url "projects/#{repo_id}/pages/domains/#{domain}"
            req.body = {
              certificate:  certificate,
              key:          key
            }.to_json
          end
          response.success?
        end

        def create_branch!
          Jekyll.logger.info "Creating branch #{branch}.."
          connection.post do |req|
            req.url  "projects/#{repo_id}/repository/branches"
            req.body = {
              branch:  branch,
              ref:     'master'
            }.to_json
          end
        end

        def commit_file!
          Jekyll.logger.info "Commiting challenge file as #{filename}"
          connection.run_request(request_method_for_commit, nil, nil, nil) do |req|
            req.url        "projects/#{repo_id}/repository/files/#{enc_filename}"
            req.body = {
              commit_message: commit_message,
              branch:         branch,
              content:        content
            }.to_json
          end
          Jekyll.logger.info "Done Commiting! Check #{gitlab_url}/#{gitlab_repo}/commits/#{branch}"
        end

      private

        def branch_exists?
          response = connection.get "projects/#{repo_id}/repository/branches"
          JSON.parse(response.body).any? { |json| json['name'] == branch }
        end

        def request_method_for_commit
          response = connection.get "projects/#{repo_id}/repository/files/#{enc_filename}?ref=#{branch}"
          response.status == 404 ? :post : :put
        end

        def enc_filename
          filename.gsub "/", "%2f"
        end

        def repo_id
          @repo_id ||= begin
            repo_name = gitlab_repo.gsub "/", "%2f"
            response  = connection.get "projects/#{repo_name}"
            unless response.success?
              fail StandardError, "Failed response for projects/#{repo_name}. Please check if personal token and repo name are correct"
            end
            JSON.parse(response.body)['id']
          end
        end

        def connection
          @connection ||= Faraday.new(url: "#{gitlab_url}/api/v4/") do |faraday|
            faraday.adapter Faraday.default_adapter
            faraday.headers['Content-Type']  = 'application/json'
            faraday.headers['PRIVATE-TOKEN'] = personal_access_token
          end
        end
      end
    end
  end
end


lib/jekyll/gitlab/letsencrypt/configuration.rb
require "active_support/core_ext/object/blank"

module Jekyll
  module Gitlab
    module Letsencrypt
      class Configuration

        DEFAULT_FILENAME        = 'letsencrypt_challenge.html'
        DEFAULT_ENDPOINT        = 'https://acme-v01.api.letsencrypt.org/'
        DEFAULT_BRANCH          = 'master'
        DEFAULT_LAYOUT          = 'null'
        DEFAULT_INITIAL_DELAY   = 120
        DEFAULT_DELAY_TIME      = 15
        DEFAULT_SCHEME          = 'http'
        DEFAULT_GITLAB_URL      = 'https://gitlab.com'
        DEFAULT_COMMIT_MESSAGE  = "Automated Let's Encrypt renewal"

        REQUIRED_KEYS = %w{gitlab_repo email domain}

        class << self

          def valid?
            REQUIRED_KEYS.all? { |key| jekyll_config.has_key? key } && personal_access_token
          end

          def endpoint
            jekyll_config['endpoint'] || DEFAULT_ENDPOINT
          end

          def gitlab_url
            jekyll_config['gitlab_url'] || DEFAULT_GITLAB_URL
          end

          def gitlab_repo
            jekyll_config['gitlab_repo']
          end

          def base_path
            jekyll_config['base_path'] || ''
          end

          def pretty_url?
            !!jekyll_config['pretty_url']
          end

          def append_str
            jekyll_config['append_str'] || ''
          end

          def layout
            jekyll_config['layout'] || DEFAULT_LAYOUT
          end

          def personal_access_token
            jekyll_config['personal_access_token'].presence || ENV['GITLAB_TOKEN'].presence
          end

          def email
            jekyll_config['email']
          end

          def domain
            jekyll_config['domain']
          end

          def branch
            jekyll_config['branch'] || DEFAULT_BRANCH
          end

          def filename
            jekyll_config['filename'] || DEFAULT_FILENAME
          end

          def initial_delay
            jekyll_config['initial_delay'] || DEFAULT_INITIAL_DELAY
          end

          def delay_time
            jekyll_config['delay_time'] || DEFAULT_DELAY_TIME
          end

          def scheme
            jekyll_config['scheme'] || DEFAULT_SCHEME
          end

          def reset!
            @jekyll_config = nil
          end

          def jekyll_config
            @jekyll_config ||= (Jekyll.configuration({})['gitlab-letsencrypt'] || {})
          end

          def commit_message
            jekyll_config['commit_message'] || DEFAULT_COMMIT_MESSAGE
          end
        end
      end
    end
  end
end
